### DISCUSSION TITLE
What happens to diagnostics sent by the LS

### DISCUSSION THREAD

**Discussion by mdbergmann** [joaotavora/eglot#683] [General]
I'm working with eglot on a Scala project using Metals.
This project, due to Scapegoat compiler plugin, produces tons of diagnostics.
It seems that Metals passes through all diagnostics of the complete project to the LSP client.
How are they handled by the eglot? Are they stored on disk, like a cache?
Assumingly they are passed over to flymake.
It is problematic when there are many diagnostics?
Is this actually the right use of the protocol, when diagnostics of the complete project are sent to the client even though only a tiny fraction are used in the open files?

**Comment by joaotavora** [comment]
> How are they handled by the eglot? Are they stored on disk, like a cache?

No, there is no cache.

> Is this actually the right use of the protocol, when diagnostics of the complete project are sent to the client even though only a tiny fraction are used in the open files?

It is a legal use of the protocol but Eglot, and mostly Flymake doesn't know what to do with that extra information right now.  That could change, and I think it's a feature for Flymake. I have some ideas, but I've yet to find the manpower (which is mostly me) to implement it in a good architectural way.  There is an issue about this: https://github.com/joaotavora/eglot/issues/504 and a pull request https://github.com/joaotavora/eglot/pull/640.  I'm going to close the issue in favour of this discussion.

Before even touching the code, one needs to understand, under various clearly laid out scenarios, what should happen in terms of --

a) in-buffer annotation (i.e. the in-buffer underlines)
b)  diagnostic summary under `M-x flymake-diagnostics-buffer`

in an hypothetical project with files `foo.c`, `bar.c`, `baz.c`, `quuz.c` when:

1) LSP sends diagnostics for all of them but only `foo.c` and `bar.c` are being visited by buffers.
2) LSP sends diagnostic for a subset of these files
3) Same as 1 but there were already some diagnostics previously recorded for "unvisited" files `baz.c` and `quuz.c`
4) Probably more that I'm not thinking of.

It's only when the combination of these scenarios is very clearly laid out and understood that we should begin coding.  Most of this work should be done in Flymake.  Flymake lives in the Emacs repo.  Flymake has automatic tests.  It would probably be a good beginning to begin speccing this in the automatic tests of Flymake, but it can maybe also be done in the automatic tests of Eglot.

**Reply by mdbergmann** [reply]
I've asked this also because I'm trying to investigate issues I have with Metals (Scala LS).
When Eglot makes connection to Metals it sends endless streams of diagnostics for mostly unvisited files but also others. There are a lot. Sometimes this takes minutes. This is probably because of a linting Scala compiler plugin which produces a lot of diagnostics.
Overall this setup is very unstable, with timeouts for requests to the server, etc. where I have to terminate java processes to eventually get to a usable state. It's unfortunately difficult to find the reason for that without more knowledge on the server side.

However, that brought me to: why are all the full diagnostics sent to the client?
Some clients show those in a file explorer or, as you say, some summary screen like what flymake could do.
But yet. Isn't a 'summary' not enough until someone really either a) opens the files and requires full diagnostics, or b) drills down to the files in the summary view which might be on a file basis.
This high-level summary could just be: file X has x infos, y warnings and z errors. What they exactly are is maybe not important at this stage. Full diagnostics can be loaded on demand.

**Reply by mdbergmann** [reply]
As a side-note. I would love to get a bit involved with development on Emacs.

What I mentioned about this 'summary' info. I'm not sure something like this actually exists in LSP.
Or would it be necessary to really transfer all, or almost all diagnostics to the client?
It depends I guess if this whole thing is transforming towards a full IDE kind of thing.
But, even IntelliJ IDEA, which I used for the last 12 years and still use, does not really have that grand overview of diagnostics for everything. Eventually you have to build the full project to see where things still do not compile.

**Reply by joaotavora** [reply]
I think this has to be broken down piece-wise.  For now, I'd like to tackle the possibility of presentation of diagnostics for multiple files in Flymake.

Then we can look at how to make Eglot's processing of enormous amounts of diagnostic info coming from the LSP server more efficient.  

These are two different problems.

**Reply by mdbergmann** [reply]
Sure. No prob.

For my immediate problem:

> Then we can look at how to make Eglot's processing of enormous amounts of diagnostic info coming from the LSP server more efficient.

How does Eglot currently handle this large amount of data.
Is it just dropped? Or is it kept in memory?

**Reply by joaotavora** [reply]
> How does Eglot currently handle this large amount of data.
Is it just dropped? Or is it kept in memory?

The parts relevant to files that are visited by buffers that Eglot happens to be managing are converted into Flymake diagnostics and serve to annotate the buffer and fill the diagnostic summary list.  The other parts are discarded.  But to process a big JSON object you have to read it to the end, so at one point it will be all in memory.  Worse the big string will be in memory as a string and as Elisp objects.

One promising possibility of improvement is developing a lazy JSON parser that converts the string to Elisp objects "just in time".

**Reply by dillon-0** [reply]
Sorry If I'm being repetitive, or if this is no longer relevant, but I'm wondering if `flymake-show-project-diagnostics`  could work with `eglot` to show diagnostics for all files (open or not) in the project. Any thoughts on the difficulty or usefulness of this @joaotavora ?

**Reply by joaotavora** [reply]
It's useful, but not 100% within the control of Eglot.  Last I checked, some servers allow this, some servers don't.  It depends on whether the server actually sends this information.
